### PR TITLE
Add notes feature, init UI Testing

### LINKED ISSUE #64: As a user, I want to be able to add notes to people to keep track of certain things

### PR DESCRIPTION
**Pull request by conradsoon**
This fixes #64

### CODE REVIEW

File: src/main/java/seedu/address/logic/parser/NoteCommandParser.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+    /**
+     * Parses the given {@code String} of arguments in the context of the NoteCommand
+     * and returns a NoteCommand object for execution.
+     * @throws ParseException if the user input does not conform to the expected format
+     */
+    public NoteCommand parse(String args) throws ParseException {
+        try {
+            // Split based on space
+            String[] splitArgs = args.trim().split("\\s", 2);
```
Comment (lordidiot) [inline src/main/java/seedu/address/logic/parser/NoteCommandParser.java:23]: TIL about the second arg

File: src/main/java/seedu/address/model/person/Person.java [hunk 2]
```diff
@@ -154,6 +160,25 @@ public boolean isSamePerson(Person otherPerson) {
...
+    public void addNote(Note note) {
+        notes.add(note);
+    }
+
+    /**
+     * Removes a note from the person.
+     * @param idx
+     * @throws Exception
+     */
+    public void removeNote(int idx) { //throws Exception {
```
Comment (lordidiot) [inline src/main/java/seedu/address/model/person/Person.java]: Leftover comments

File: src/main/java/seedu/address/model/person/Person.java [hunk 3]
```diff
@@ -47,14 +51,15 @@ public Person(Name name, Phone phone, Email email, Address address, Set<Tag> tag
         this.telegram = Optional.empty();
         this.tags.addAll(tags);
         this.id = Optional.empty();
+        this.notes = new LinkedList<>();
```
Comment (lordidiot) [inline src/main/java/seedu/address/model/person/Person.java]: Is there a reason for using LinkedList rather than ArrayList?

File: src/main/java/seedu/address/model/util/SampleDataUtil.java [hunk 4]
```diff
@@ -68,4 +73,13 @@ public static Set<Tag> getTagSet(String... strings) {
                 .collect(Collectors.toSet());
     }
 
+    /**
```
Comment (lordidiot) [inline src/main/java/seedu/address/model/util/SampleDataUtil.java:76]: Would be better for cohesion if getNoteList was in its own class, like `NotesUtil::getNoteList`

File: src/main/java/seedu/address/ui/PersonCard.java [hunk 5]
```diff
@@ -75,5 +80,37 @@ public PersonCard(Person person, int displayedIndex) {
...
+        int numberOfNotes = person.getNotes().size();
+        notesButton.setText("Notes (" + numberOfNotes + ")");
+
+    }
+
+    /**
+     * Opens a new window to display the notes of the person.
+     */
+    @FXML
+    public void handleNotesButtonClick() {
```
Comment (lordidiot) [inline src/main/java/seedu/address/ui/PersonCard.java:94]: What's going on here, can you explain the issue you are facing?

**Review by lordidiot — CHANGES_REQUESTED** [review]
Overall looks good, thanks for working on this! I like the data model for notes, much simpler than what I had in mind and easy to store to disk.

Noted some things that I'd like to discuss with you before finalising this PR. Also, please ensure that test coverage is non-negative before I can give an approving review.

File: src/main/java/seedu/address/ui/PersonCard.java (hunk 5 above)
Comment (conradsoon) [reply src/main/java/seedu/address/ui/PersonCard.java:94]: Basically, I'm unable to get JavaFX to create a new window displaying the notes (using the classes that have been defined in the project causes errors when I construct them). I've hacked around it by using an Alert, but if we want any more logic beyond just displaying text, we'll have to fix this at some point.

File: src/main/java/seedu/address/model/util/SampleDataUtil.java (hunk 4 above)
Comment (conradsoon) [reply src/main/java/seedu/address/model/util/SampleDataUtil.java:76]: IMHO I don't think there's enough test data methods specifically for making notes that I'll need my own class for it. Likewise for the other sample data methods actually.

I prefer this approach of putting all the "test data util" methods into the SampleDataUtil class we have now. Thoughts?

File: src/main/java/seedu/address/model/person/Person.java (hunk 3 above)
Comment (conradsoon) [reply src/main/java/seedu/address/model/person/Person.java]: No real reason, I can change it to ArrayList.

File: src/main/java/seedu/address/ui/PersonCard.java (hunk 5 above)
Comment (lordidiot) [reply src/main/java/seedu/address/ui/PersonCard.java:94]: I've traced the issue, it's an issue with your FXML structure.
@conradsoon Let's fix this and use this approach instead of the hacky alert box method.

```diff
--- a/src/main/resources/view/NotesWindow.fxml
+++ b/src/main/resources/view/NotesWindow.fxml
@@ -7,16 +7,25 @@
 <?import javafx.scene.control.Button?>
 <?import javafx.scene.control.Label?>
 <?import javafx.geometry.Insets?>
-<VBox xmlns="http://javafx.com/javafx/17" xmlns:fx="http://javafx.com/fxml/1" spacing="10" alignment="CENTER">
-    <padding>
-        <Insets top="10" right="10" bottom="10" left="10" />
-    </padding>

-    <Label text="Notes" style="-fx-font-weight: bold; -fx-font-size: 18px;" />
+<?import javafx.scene.Scene?>
+<fx:root resizable="false" title="Notes"
+         type="javafx.stage.Stage" xmlns="http://javafx.com/javafx/17" xmlns:fx="http://javafx.com/fxml/1">
+    <scene>
+        <Scene>
+            <VBox spacing="10" alignment="CENTER">
+                <padding>
+                    <Insets top="10" right="10" bottom="10" left="10" />
+                </padding>

-    <ScrollPane fitToWidth="true" prefHeight="300" VBox.vgrow="ALWAYS">
-        <ListView fx:id="notesListView" />
-    </ScrollPane>
+                <Label text="Notes" style="-fx-font-weight: bold; -fx-font-size: 18px;" />

-    <Button text="Close" />
-</VBox>
+                <ScrollPane fitToWidth="true" prefHeight="300" VBox.vgrow="ALWAYS">
+                    <ListView fx:id="notesListView" />
+                </ScrollPane>
+
+                <Button text="Close" />
+            </VBox>
+        </Scene>
+    </scene>
+</fx:root>
```

File: src/main/java/seedu/address/model/util/SampleDataUtil.java (hunk 4 above)
Comment (lordidiot) [reply src/main/java/seedu/address/model/util/SampleDataUtil.java:76]: Sure, let's keep it as is

**Review by kokrui — APPROVED** [review]
LGTM! Thanks for taking the step to begin automated UI testing from our repo, this will help us a ton!

File: src/main/java/seedu/address/ui/PersonCard.java [hunk 6]
```diff
@@ -78,5 +82,21 @@ public PersonCard(Person person, int displayedIndex) {
...
+        try {
```
Comment (lordidiot) [inline src/main/java/seedu/address/ui/PersonCard.java:95]: Is this try catch necessary?

**Review by lordidiot — APPROVED** [review]
Looks great! Thanks for making the changes.
